src/uri/uri_parser.cr

```diff
@@ -0,0 +1,238 @@
+# Parser based on https://url.spec.whatwg.org/
+class URIParser
+  property uri
+
+  macro cor(method)
+    return {{method}}
+  end
```

Comment:
I should have a better name for it, but it allows better testing in the specs without having to have extra conditionals in a real program. The tests can work on a single step, and assert what the next step will be, instead of having the next steps all just called.